### PR TITLE
[TRTLLM-11412][fix] Preserve dense tensor layouts during offloading

### PR DESCRIPTION
**Pull request by rahul-steiger-nv**
## Description

Fixes [NVBug 6702264](https://nvbugspro.nvidia.com/bug/6702264).

Allow the offload manager to pack non-overlapping dense tensors such as Wan VAE
`channels_last_3d` weights. Tensor bytes are copied in physical storage order so
the original strides and values are preserved when CPU and GPU views are rebound.
Overlapping and non-dense layouts remain unsupported.

## Dev Engineer Review

- Offload packing now supports non-overlapping dense tensors, including `channels_last_3d` layouts.
- Physical storage-order copying preserves values and strides during CPU and GPU rebinding.
- Overlapping and non-dense layouts remain unsupported.
- No public API or configuration changes are reported.

## QA Engineer Review

- Modified `tests/unittest/_torch/visual_gen/test_offloading.py`.
- Added coverage for `channels_last_3d` value and stride preservation.
- Added rejection coverage for overlapping and non-dense layouts.
- The test file was not found in the inspected `test-db/` or `qa/` lists.
- The GitHub PR CI run succeeded. The L0 merge-request pipeline was unstable.
- Multi-GPU validation requires the `ci: full pre-merge approved` label and a CI re-run.
- Verdict: needs follow-up.

## Per-File QA Perspective

- `tensorrt_llm/_torch/visual_gen/offloading.py`: Verify dense tensors preserve values and strides during CPU and GPU staging. Verify unsupported layouts remain rejected.
- `tests/unittest/_torch/visual_gen/test_offloading.py`: Covers `channels_last_3d` preservation and invalid-layout rejection. The test was not found in the inspected CI or manual-QA lists.

## Test Coverage

- Added `test_offload_preserves_channels_last_3d_layout` to verify values and
strides after initialization and staging.

## PR Checklist

Please review the following before submitting your PR:

- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/offloading.py [hunk 1]
```diff
@@ -179,9 +191,9 @@ def _build_spec(
         offset: int,
     ) -> _FlatTensorSpec:
         display_name = f"{stage_name}.{qualified_name}"
-        if not tensor.is_contiguous():
+        if not self._is_non_overlapping_and_dense(tensor):
             raise ValueError(
-                f"Cannot offload non-contiguous tensor '{display_name}' "
+                f"Cannot offload overlapping or non-dense tensor '{display_name}' "
                 f"with stride {tuple(tensor.stride())}"
             )
```
Comment (karljang) [inline tensorrt_llm/_torch/visual_gen/offloading.py:198]: Non-blocking: this guard is being loosened, and neither the old `is_contiguous()` branch nor the new one has a negative test - `test_offloading.py` currently only exercises `ValueError` for unknown stage names. Since the correctness of the whole packing scheme now rests on this predicate, it would be worth pinning a case that must still be rejected, e.g.:

```python
class _OverlappingToyModule(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.register_buffer("broadcast", torch.zeros(1, 3).expand(4, 3))


def test_offload_rejects_overlapping_tensor():
    manager = ModuleOffloadManager(
        stages={"stage": _OverlappingToyModule()}, device="cpu", pin_memory=False
    )
    with pytest.raises(ValueError, match=r"overlapping or non-dense"):
        manager.initialize()
```

A non-dense case such as `torch.zeros(4, 6)[:, ::2]` would cover the other half of the predicate.

**Review by karljang — APPROVED** [review]
LGTM. The three pieces line up:

- `_is_non_overlapping_and_dense` is the standard predicate (sort `(stride, size)`, skip size<=1 dims, require the recurrence `1, n0, n0*n1, ...`). It admits any dense permutation and still rejects `expand`-style overlap and strided gaps. I hand-checked it against Conv3d weights with size-1 kernel dims, e.g. `(96, 96, 1, 3, 3)` -> strides `(864, 1, 864, 288, 96)`, where the size-1 dim's stride is correctly ignored.
- `as_strided((numel,), (1,))` copies the dense span from the tensor's own `storage_offset` (`as_strided` inherits it when `storage_offset=None`), which is exactly `spec.nbytes`. This is the real fix: `reshape(-1)` materialized a copy in *logical* order, which then got reinterpreted with the physical stride at rebind time.
- `_make_views` restores `spec.stride` at the packed offset, and `stage()` is a raw byte copy of the whole stage buffer, so the layout survives the GPU transfer too.

Aliasing still behaves: two permuted views of one storage copy identical raw bytes, and each is rebound with its own stride.

One minor, non-blocking comment inline about the untested rejection branch.

Two notes:

1. The auto-generated summary in the description says "No `tests/integration/test_lists/` coverage entry was identified. Verdict: needs follow-up." That looks incorrect - `tests/integration/test_lists/test-db/l0_b200.yml:286` lists the whole file `unittest/_torch/visual_gen/test_offloading.py`, so `test_offload_preserves_channels_last_3d_layout` is collected in l0_b200. No follow-up needed there.
2. Nit on the new docstring: the tensor densely covers a contiguous span starting at its `storage_offset`, not "its storage" - a slice of a larger allocation passes the check, and should.

Disclosure: I reviewed by reading `as_strided`/`narrow`/`view` semantics and hand-evaluating the stride recurrence; I did not execute the new test locally.

File: tensorrt_llm/_torch/visual_gen/offloading.py (hunk 1 above)
Comment (rahul-steiger-nv) [reply tensorrt_llm/_torch/visual_gen/offloading.py:198]: thanks for the review, I updated it

File: tensorrt_llm/_torch/visual_gen/offloading.py [hunk 2]
```diff
@@ -142,6 +142,18 @@ def _owner_and_name(root: nn.Module, qualified_name: str) -> tuple[nn.Module, st
     def _tensor_nbytes(tensor: torch.Tensor) -> int:
         return tensor.numel() * tensor.element_size()
 
+    @staticmethod
+    def _is_non_overlapping_and_dense(tensor: torch.Tensor) -> bool:
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/offloading.py:147]: why not directly using pytorch internal helper directly?
```
from torch._prims_common import is_non_overlapping_and_dense_or_false

@staticmethod
def _is_non_overlapping_and_dense(tensor: torch.Tensor) -> bool:
    return is_non_overlapping_and_dense_or_false(tensor)
```

Comment (rahul-steiger-nv) [reply tensorrt_llm/_torch/visual_gen/offloading.py:147]: updated, thanks!

**Review by zhenhuaw-me — APPROVED** [review]

**Review by chang-l — APPROVED** [review]
